File: database_project_1/Scripts/src/OptimizedImporter.java
import java.nio.file.Path;
import java.sql.*;
import java.util.List;

public class OptimizedImporter {
    public static void main(String[] args) throws SQLException {
        long startTime = System.nanoTime();

        try {
            // Establish a connection to your PostgreSQL database
            Connection conn = DriverManager.getConnection("jdbc:postgresql://localhost:5432/cs307", "harroldtok", "6969");
            conn.setAutoCommit(false);

            // JSON data obtained from your file parsing logic
            //List<Cards> cards = ReadJSON.readJsonArray(Path.of("resource/cards.json"), Cards.class);
            List<Ride> rides = ReadJSON.readJsonArray(Path.of("resource/ride.json"), Ride.class);
           // String sql = "INSERT INTO lab3.cards (code, money, create_time) VALUES (?, ?, ?)";

            String sql1 = "INSERT INTO test.cardid_rides (user_code, start_station, end_station, price, start_time, end_time) VALUES (?, ?, ?, ?, ?, ?)";
            String sql2 = "INSERT INTO test.userid_rides (user_id, start_station, end_station, price, start_time, end_time) VALUES (?, ?, ?, ?, ?, ?)";
            PreparedStatement pstmt1 = conn.prepareStatement(sql1);
            PreparedStatement pstmt2 = conn.prepareStatement(sql2);
           // PreparedStatement pstmt = conn.prepareStatement(sql);

            //test optimized script on cards
//            for (Cards c: cards) {
//                pstmt.setString(1, c.getCode());
//                pstmt.setString(2, String.valueOf(c.getMoney()));
//                pstmt.setString(3, String.valueOf(c.getCreate_time()));
//            }
            for (Ride r: rides) {
                if(r.getUser().length() == 9){
                    pstmt1.setString(1, r.getUser());
                    pstmt1.setString(2, r.getStartStation());
                    pstmt1.setString(3, r.getEndStation());
                    pstmt1.setDouble(4, r.getPrice());
                    pstmt1.setTimestamp(5, r.getStartTime());
                    pstmt1.setTimestamp(6, r.getEndTime());
                    pstmt1.addBatch();
                } else{
                    pstmt2.setString(1, r.getUser());
                    pstmt2.setString(2, r.getStartStation());
                    pstmt2.setString(3, r.getEndStation());
                    pstmt2.setDouble(4, r.getPrice());
                    pstmt2.setTimestamp(5, r.getStartTime());
                    pstmt2.setTimestamp(6, r.getEndTime());
                    pstmt2.addBatch();
                }
            }
            pstmt1.executeBatch();
            pstmt2.executeBatch();
            conn.commit();
            conn.setAutoCommit(true);
            pstmt1.close();
            pstmt2.close();
        } catch (SQLException e) {
            e.printStackTrace();
        }
        long endTime = System.nanoTime();
        long elapsedTime = endTime - startTime;
        double elapsedTimeSeconds = (double) elapsedTime / 1_000_000_000.0;
        System.out.println("Elapsed Time: " + elapsedTimeSeconds + " seconds");
    }
}

File: database_project_1/PythonScripts/OptimizedImporter.py
import psycopg2
from psycopg2.extras import execute_values  # Import the correct function
import json
import time

# Define connection parameters
DB_HOST = 'localhost'
DB_PORT = '5432'
DB_NAME = 'cs307'
DB_USER = 'harroldtok'
DB_PASSWORD = '6969'

# Load JSON data (you can replace the sample data with actual JSON data)
with open('/Users/harroldtok/PycharmProjects/DataImport/ride.json') as f:
    rides = json.load(f)  # Ensure the JSON file is correctly formatted

# Start timer to measure execution time
start_time = time.time()

# Establish a connection to the PostgreSQL database
conn = psycopg2.connect(
    host='127.0.0.1',
    port=5432,
    dbname='cs307',
    user='harroldtok',
    password=6969
)

# Start a transaction block
conn.autocommit = False
try:
    # Prepare SQL queries
    sql1 = "INSERT INTO lab3.cardid_rides (user_code, start_station, end_station, price, start_time, end_time) VALUES %s"
    sql2 = "INSERT INTO lab3.userid_rides (user_id, start_station, end_station, price, start_time, end_time) VALUES %s"

    # Create prepared statements
    pstmt1 = conn.cursor()
    pstmt2 = conn.cursor()
    cursor = conn.cursor()

    cardid_rides = []
    userid_rides = []
    # Add batch insertions
    for r in rides:
        record = (r['user'], r['start_station'], r['end_station'], r['price'], r['start_time'], r['end_time'])
        if len(r['user']) == 9:
            cardid_rides.append(record)
        else:
            userid_rides.append(record)

        # Execute batch operations
    if cardid_rides:
        execute_values(cursor, sql1, cardid_rides)

    if userid_rides:
        execute_values(cursor, sql2, userid_rides)

        # Commit the transaction after successful batch execution
    conn.commit()

    # Close the prepared statements
    pstmt1.close()
    pstmt2.close()

except psycopg2.Error as e:
    # Rollback in case of errors
    print("An error occurred:", e)
    conn.rollback()

finally:
    # Ensure the connection is closed
    conn.close()

# Calculate elapsed time
end_time = time.time()
elapsed_time_seconds = end_time - start_time
print(f"Elapsed Time: {elapsed_time_seconds:.2f} seconds")
